File: frameworks/go/gin/Implementation/stream.go
package implementation

import (
	"io"

	"github.com/gin-gonic/gin"
	ginjson "github.com/gin-gonic/gin/codec/json"
)

// streamRoutes write items.medium's rows one per line through c.Stream, which flushes after each
// step. The length is never known, so the answer goes out chunked.
func streamRoutes(r *gin.Engine, p *Payloads) {
	r.GET("/stream/items", func(c *gin.Context) {
		rows := p.Medium.Items
		c.Header("Content-Type", "application/x-ndjson")
		next := 0
		c.Stream(func(w io.Writer) bool {
			// The encoder ends each row with a newline.
			if err := ginjson.API.NewEncoder(w).Encode(&rows[next]); err != nil {
				return false
			}
			next++
			return next < len(rows)
		})
	})
}

File: frameworks/rust/axum/Implementation/routes/stream.rs
use axum::body::{Body, Bytes};
use axum::http::header::CONTENT_TYPE;
use axum::routing::get;
use axum::Router;
use futures_util::stream::{self, StreamExt};

use crate::Payloads;

/// stream: items.medium's rows written one per line, each line a chunk of its own. The body is a
/// stream, so its length is never known and the answer goes out chunked.
pub fn router(p: &'static Payloads) -> Router {
    Router::new().route("/stream/items", get(move || async move {
        let lines = stream::iter(&p.medium.items).map(|row| {
            serde_json::to_vec(row).map(|mut line| {
                line.push(b'\n');
                Bytes::from(line)
            })
        });
        ([(CONTENT_TYPE, "application/x-ndjson")], Body::from_stream(lines))
    }))
}
